source/common/config/api_type_oracle.cc

```diff
@@ -31,5 +31,20 @@ ApiTypeOracle::getEarlierVersionMessageTypeName(const std::string& message_type)
   }
   return absl::nullopt;
 }
+
+const absl::optional<std::string> ApiTypeOracle::getEarlierTypeUrl(const std::string& type_url) {
+  size_t type_name_start = type_url.find('/');
+  if (type_name_start == std::string::npos) {
+    return {};
+  }
+  std::string message_type = type_url.substr(type_name_start + 1);
+  absl::optional<std::string> old_message_type =
+      ApiTypeOracle::getEarlierVersionMessageTypeName(message_type);
+  if (old_message_type.has_value()) {
+    return "type.googleapis.com/" + old_message_type.value();
```

Comment:
You probably want to generally work with type names in this file, and only at the point at which you are doing type URLs in the watcher etc. do the conversion to type URL. FWIW, I think we should have a `TypeUtil::descriptionFullNameToTypeurl` to factor out this low level manipulation.

Reply:
I have moved this helper function to api_type_oracle.cc and factor out the low level manipulation.